### ISSUE TITLE
Open task for feature requests

### ISSUE THREAD

**Issue by johnkerl** [johnkerl/miller#77]
Feel free to open a new task, or just add discussion as a comment here.

**Comment by jungle-boogie** [comment]
Hi,

This is mostly thinking out loud...

I very much like your tool because it does so much so quickly. There are many types of tools in the data processing field but they rely on higher level languages and, I'm guessing, as a consequence, they run a bit slower for me. I have a 7.5MB 80,000 line file and it can compute a very simple request (just `count-distinct then sort -nr count` on postal codes) in 3/4 of a second with 42,314 lines.

Other utilities you may be interested in:
https://github.com/dbohdan/sqawk
https://github.com/sordina/SQLBong
https://github.com/harelba/q
And so on.

Have you considered something like `sort -flag limit 20`? What about json output or a table output?

thanks!

**Comment by johnkerl** [comment]
For `sort -flag limit 20` do you mean for example (using 5 or 2 instead of 20 to save screen space here):

```
$ cat ../data/big.dkvp | mlr sort -nr x then head -n 5
a=zee,b=wye,i=853631,x=0.999999276362359191,y=0.987837822170852586
a=zee,b=eks,i=312012,x=0.999999214429407979,y=0.986796915059348967
a=hat,b=pan,i=616674,x=0.999997778329997256,y=0.962660392263280373
a=eks,b=hat,i=889555,x=0.999997151549904206,y=0.952126099240093504
a=eks,b=hat,i=586109,x=0.999995659571139006,y=0.927050412132893853
```

```
$ cat ../data/big.dkvp | mlr top -a -n 5 -f x
a=zee,b=wye,i=853631,x=0.999999276362359191,y=0.987837822170852586
a=zee,b=eks,i=312012,x=0.999999214429407979,y=0.986796915059348967
a=hat,b=pan,i=616674,x=0.999997778329997256,y=0.962660392263280373
a=eks,b=hat,i=889555,x=0.999997151549904206,y=0.952126099240093504
a=eks,b=hat,i=586109,x=0.999995659571139006,y=0.927050412132893853
```

```
$ cat ../data/big.dkvp | mlr top -a -n 2 -f x -g a
a=wye,b=wye,i=835662,x=0.999991802964355747,y=0.862232421926330983
a=wye,b=zee,i=558352,x=0.999986721668386203,y=0.776831080567478738
a=eks,b=hat,i=889555,x=0.999997151549904206,y=0.952126099240093504
a=eks,b=hat,i=586109,x=0.999995659571139006,y=0.927050412132893853
a=zee,b=wye,i=853631,x=0.999999276362359191,y=0.987837822170852586
a=zee,b=eks,i=312012,x=0.999999214429407979,y=0.986796915059348967
a=pan,b=pan,i=444121,x=0.999994222540405708,y=0.902898236598306481
a=pan,b=hat,i=743855,x=0.999987163580948080,y=0.784258304994673638
a=hat,b=pan,i=616674,x=0.999997778329997256,y=0.962660392263280373
a=hat,b=hat,i=60978,x=0.999995048623529725,y=0.916782215664527445
```

The idea is you can run the `sort` output through `head` to keep the desired number of records. But that runs longer than needed: it sorts all records and then throws away all but the top few. So you can instead use Miller's `top` which only every retains in memory the top n (e.g. 20) records at any one time. Same output between the first and second example just above. The third example is a reminder that `top` works with category fields, too, if you want.

**Comment by johnkerl** [comment]
For "table output" do you mean for example
```
% cat ../data/big.dkvp | mlr --opprint top -a -n 2 -f x -g a
a   b   i      x                    y
wye wye 835662 0.999991802964355747 0.862232421926330983
wye zee 558352 0.999986721668386203 0.776831080567478738
eks hat 889555 0.999997151549904206 0.952126099240093504
eks hat 586109 0.999995659571139006 0.927050412132893853
zee wye 853631 0.999999276362359191 0.987837822170852586
zee eks 312012 0.999999214429407979 0.986796915059348967
pan pan 444121 0.999994222540405708 0.902898236598306481
pan hat 743855 0.999987163580948080 0.784258304994673638
hat pan 616674 0.999997778329997256 0.962660392263280373
hat hat 60978  0.999995048623529725 0.916782215664527445
```

This is what the pprint format is for.

**Comment by johnkerl** [comment]
Regarding JSON, that's another story, and one deserving of a separate write-up. Miller handles tabular data -- iterables of insertion-ordered hashmaps -- lists of records which in turn are lists of fields which are key-value pairs -- these are all various ways to say the same thing. JSON (and XML for that matter) is a far, far richer format -- supporting list of maps containing lists of lists of lists or what have you. JSON and XML are *recursive data structures* and a tool like Miller is never going to go very far down that road.

One answer is to point you at https://stedolan.github.io/jq/.

Another is that Miller could handle JSON or XML, *as long as* it's understood that it'll only take lists of maps (i.e. tabular data) and nothing richer. And within that scope, there are two ways to do it: one is JSON support within Miller, and another is pre/post-filtering with something like `sed`.

To be very specific, a little CSV data such as
```
a,x
zee,3.1
wye,7.2
```
could be represented in DKVP by
```
a=zee,x=3.1
a=wye,x=7.2
```
and in JSON by
```
[
  {
    "a": "zee",
    "x": "3.1"
  },
  {
    "a": "wye",
    "x": "7.2"
  },
]
```
It would be just some regular-expression work to convert JSON to DKVP: remove the square brackets and double-quotes, treat the curly braces as record-delimiters, use colon as pair-separator, and so on.

Again, I'll write this up in more detail sometime soon, but for now that's a sketch.

**Comment by jungle-boogie** [comment]
Hi @johnkerl,

**RE: top vs. head**
Yes, your examples are perfect. I had not known either head or top were builtin. I was piping through the OS head, which made the results much longer than either one of your solutions.
In my case, `mlr --csv --rs lf count-distinct -f ZIPCode then top -a -n 20 -f count 5-digitCommercial.csv` was slightly more faster than miller's head option.

**RE: table output**
opprint does format very nicely but I was asking about an actual table around the data. Like a table drawn with --- ||| around the data. 

**RE: json output**
Thanks for your thoughts!

**Comment by Komosa** [comment]
@jungle-boogie for nicely formatted ASCII-table you may use https://github.com/onyxfish/csvkit . It is written in python, but you are not going to display too much rows, right? :)

As a side note: I really like miller, mostly because it is maintained well :+1:

**Comment by jungle-boogie** [comment]
Hi @Komosa, yes, csvkit is what I was thinking about with ascii tables.
Yes, you're right, you won't display a table around dozens of rows

**Comment by Komosa** [comment]
General question: what do you do when some fields are unavailable (ie. rows are not homogenous)?
I encounter this problem with `stats1`.

**Comment by johnkerl** [comment]
Miller intends to handle heterogeneity as a primary feature. I missed a few spots such as stats1 which are fixed in head & will be included in the next release.

**Comment by jpc4242** [comment]
Support for input compressed files (I'm thinking ,gz, others probably also apply). 
This would allow to manage directories with hundreds of .csv.gz files by using "mlr cat". Also very handy for join

**Comment by johnkerl** [comment]
True. `zcat ... | mlr ...` is a poor-man's substitute.

**Comment by jpc4242** [comment]
(Before anything else: GREAT TOOL!!)

Yes, when wroking with csv if you use pipes you loose reference to "first line". I.e
    zcat 1,csv,gz 2,csv,gz  | mlr --icsvlite .....
does consider header line of file 2 as just data according to header of file 1

Workaround (maybe you want to include it in the FAQ):
   for i in *.csv.gz; do zcat $i | mlr --icsvlite cat; done | mlr -idkvp .....

My proposal would be to add an additional "compression layer" and be able to do something like:

mlr --input-compress=xz --output-compress=gz --icsvlite --onidx .......

But I just "landed" in your project so I don't know how easy that would be.

Also, I don't know if I should open an issue for the following: dots in filed names for filter

I have a CSV file with fields like "pfx.name". Trying to use that in filter gives "Syntax Error!" with every escaping/quoting I could think of. Probably "mlr rename" is a workaround, but an enhancement would be to be able to use them. For example something like in bash:
  mrl filter '${pfx.name} == "value"'

**Comment by johnkerl** [comment]
Re escaping: nice. I should have done this sooner but didn't think of it; glad you asked!

https://github.com/johnkerl/miller/commit/f4b95c439c9efe813b0c1d17cb048233a997e1a3

This will go out in the next minor release (maybe a week from now).

**Comment by johnkerl** [comment]
Re compression: I'll look around for libraries I can portably use.

**Comment by johnkerl** [comment]
Meanwhile there's an `xargs` trick which would do well in the FAQ for gzip-file handling.

**Comment by johnkerl** [comment]
P.S. thanks for the compliment -- glad it's working well for you!

**Comment by johnkerl** [comment]
2nd half of field-name escapes: https://github.com/johnkerl/miller/commit/bc1d5d966855eb38a8a269e82cfdd489f4f01c01

**Comment by johnkerl** [comment]
Compression support:
* https://github.com/johnkerl/miller/commit/63e5dda2f066f48cd565c7c2937d31945830fac5
* http://johnkerl.org/miller/doc/reference.html#Compression
* http://johnkerl.org/miller/doc/reference.html#On-line_help

**Comment by jpc4242** [comment]
csv unquote on input?
I know we have option for output quoting (always, minimal, ..), but I didn't find an option to remove quotes on the input.
Rationale: I'm trying to avoid this:
<block>
$ cat kk1
id,name,category
34,"miller, the useful tool",data parsing
$ mlr --icsvlite --ifs comma cat kk1
mlr: Header-data length mismatch in file kk1 at line 2.
</block>

**Comment by johnkerl** [comment]
`--csvlite` doesn't do quoting. It's my non-standard, ad-hoc way to do heterogeneous CSV, without the performance-impacting messiness of handling quoting, embedded commas, etc.

`--csv` is RFC-compliant CSV & handles double-quoted input & has options for what quoting behavior you want on ouput:

```
mlr -h
...
Double-quoting for CSV output:
  --quote-all        Wrap all fields in double quotes
  --quote-none       Do not wrap any fields in double quotes, even if they have
                     OFS or ORS in them
  --quote-minimal    Wrap fields in double quotes only if they have OFS or ORS
                     in them (default)
  --quote-numeric    Wrap fields in double quotes only if they have numbers
                     in them
```

and
```
$ mlr --csv --rs lf cat kk1
id,name,category
34,"miller, the useful tool",data parsing
```

**Comment by jpc4242** [comment]
MY FAULT! I forgot about the "--rs" option.
Thanks again,
   jpc

**Comment by johnkerl** [comment]
No worries. :D 

P.S. I will admit, there are a LOT of options. ;)

**Comment by aaronwolen** [comment]
This is such a great tool; I rely on it more and more every day. Since you're soliciting feature requests, I'd *love* to be able to reshape data with miller, converting between "wide" and "long" formats. For example:

Wide format
```
         time           X          Y           Z
1  2009-01-01  0.65473572  2.4520609 -1.46570942
2  2009-01-02 -0.89248112  0.2154713 -2.05357735
3  2009-01-03  0.98012375  1.3179287  4.64248357
4  2009-01-04  0.35397376  3.3765645 -0.25237774
5  2009-01-05  2.19357813  1.3477511  0.09719105
```

Long format

```
         time stock       price
1  2009-01-01     X  0.65473572
2  2009-01-02     X -0.89248112
3  2009-01-03     X  0.98012375
4  2009-01-04     X  0.35397376
5  2009-01-05     X  2.19357813
6  2009-01-01     Y  2.45206093
7  2009-01-02     Y  0.21547134
8  2009-01-03     Y  1.31792866
9  2009-01-04     Y  3.37656453
10 2009-01-05     Y  1.34775108
11 2009-01-01     Z -1.46570942
12 2009-01-02     Z -2.05357735
13 2009-01-03     Z  4.64248357
14 2009-01-04     Z -0.25237774
15 2009-01-05     Z  0.09719105
```

The [reshape](https://github.com/hadley/reshape) package for R is a great model.

**Comment by johnkerl** [comment]
awesome :)

**Comment by johnkerl** [comment]
`mlr reshape` with wide-to-long is super-easy, and is already in head (no docs yet, but soon). The long-to-wide is just a smidge more work (and is non-streaming: it will require all input records to be read before producing any output records).

**Comment by johnkerl** [comment]
```
$ cat input.tbl
time       X           Y         Z
2009-01-01 0.65473572  2.4520609 -1.46570942
2009-01-02 -0.89248112 0.2154713 -2.05357735
2009-01-03 0.98012375  1.3179287 4.64248357
2009-01-04 0.35397376  3.3765645 -0.25237774
2009-01-05 2.19357813  1.3477511 0.09719105

$ mlr --pprint reshape -i X,Y,Z -o item,price input.tbl
time       item price
2009-01-01 X    0.65473572
2009-01-01 Y    2.4520609
2009-01-01 Z    -1.46570942
2009-01-02 X    -0.89248112
2009-01-02 Y    0.2154713
2009-01-02 Z    -2.05357735
2009-01-03 X    0.98012375
2009-01-03 Y    1.3179287
2009-01-03 Z    4.64248357
2009-01-04 X    0.35397376
2009-01-04 Y    3.3765645
2009-01-04 Z    -0.25237774
2009-01-05 X    2.19357813
2009-01-05 Y    1.3477511
2009-01-05 Z    0.09719105

$ mlr --pprint reshape -r '^[A-Z]$' -o item,price input.tbl
time       item price
2009-01-01 X    0.65473572
2009-01-01 Y    2.4520609
2009-01-01 Z    -1.46570942
2009-01-02 X    -0.89248112
2009-01-02 Y    0.2154713
2009-01-02 Z    -2.05357735
2009-01-03 X    0.98012375
2009-01-03 Y    1.3179287
2009-01-03 Z    4.64248357
2009-01-04 X    0.35397376
2009-01-04 Y    3.3765645
2009-01-04 Z    -0.25237774
2009-01-05 X    2.19357813
2009-01-05 Y    1.3477511
2009-01-05 Z    0.09719105
```

**Comment by johnkerl** [comment]
OK @aaronwolen this is in head: see also http://johnkerl.org/miller/doc/reference.html#reshape

Are you a pull-from-head-and-compile person, or should I cut a tag for you to try it out?

**Comment by aaronwolen** [comment]
Wow, that was fast! Compiling isn't a problem. I'll put it through the wringer and report back.

**Comment by aaronwolen** [comment]
It works perfectly and (not surprisingly) is incredibly fast. I put together a crude benchmark with a few popular packages for data shaping and only [data.table](https://github.com/Rdatatable/data.table) could match miller's performance.

This is a killer addition to an awesome tool. Thanks!

**Comment by johnkerl** [comment]
Glad to hear it!!! :)

BTW, re cutting a release ... I've made the perhaps-mistake of biting off several distinct large pieces of work (JSON I/O, out-of-stream variables for `mlr put`, and awk-style begin/end/pattern-action blocks for `mlr put`) all of which are in some state of flux, and don't plan to cut a new release until 4.0.0 which will surely be some weeks away.

**Comment by johnkerl** [comment]
P.S. @aaronwolen you should repull from head. There is a memory-management bugfix, in reshape of which I only happened to observe symptoms in JSON format.

**Comment by aaronwolen** [comment]
What do you think about working with nested columns?

```
 x     y
 1     a
 2 d,e,f
 3   g,h
```

Where something like `mlr unnest --fs comma -f y` would produce

```
x     y
1     a
2     d
2     e
2     f
3     g
3     h
```

**Comment by johnkerl** [comment]
Yes, this is like @komosa's `explode` on https://github.com/johnkerl/miller/issues/96.

It's very easy to do *as long as* the punctuators are defined, & different from `=` and `,`. If they aren't, it's undefined where one field ends and the next one starts.

**Comment by johnkerl** [comment]
The difference between your example & @komosa's is yours replicates across multiple records, whereas his replicates across multiple fields within the same record.

**Comment by aaronwolen** [comment]
His example is somewhat complicated by the regex capturing but being able to deal with replicates in both directions would really useful. Here's a real world example I was working on today, which contains 2 different measures of association between pairs of genetic markers (`rs*`):

```
rs10	rs2237570,0.3,-0.72;rs117465896,0.75,-0.88
rs10041592	rs1628466,0.21,0.92;rs1729036,0.24,1;rs79177551,0.2,0.85
rs10033329	rs4835294,0.2,0.46
```

There are 2 fields, tab delimited. Records in the second field provide the association results for all markers relevant to the marker in field 1. Converting this format into a "long" format requires 2 operators. First, the unnesting step I described above, which spreads the semicolon-separated replicates across new records

```
rs10	    rs2237570,0.3,-0.72
rs10        rs117465896,0.75,-0.88
rs10041592	rs1628466,0.21,0.92
rs10041592  rs1729036,0.24,1
rs10041592  rs79177551,0.2,0.85
rs10033329	rs4835294,0.2,0.46
```

followed by a separating step (as described by @komosa) to spread the comma-separated replicates across new fields:

```
rs10	    rs2237570    0.3   -0.72
rs10        rs117465896  0.75  -0.88
rs10041592	rs1628466    0.21  0.92
rs10041592  rs1729036    0.24  1
rs10041592  rs79177551   0.2   0.85
rs10033329	rs4835294    0.2   0.46
```

**Comment by johnkerl** [comment]
sweet

**Comment by johnkerl** [comment]
OK looking at this more today. These are both super-simple; working on them now.

**Comment by johnkerl** [comment]
@aaronwolen and @Komosa there are explode-variants on https://github.com/johnkerl/miller/commit/9b85f1f22f81683a1dc50b84d43ffba97535d85a. I think this covers them all; please let me know if I'm wrong.

The implode-variants (undoing explode; necessarily non-streaming) are up next & will be almost as easy. This is a lot like `reshape`, as it turns out.

**Comment by aaronwolen** [comment]
Still testing it out and learning the syntax but it looks great so far. Could you clarify how the nested record/field separators should be specified? 

I see `--nested-ps` (for key-pair values, I'm assuming) and `--nested-fs` for fields but no `--nested-rs` for records?

I created a nested test filed based on your `colored-shapes.dkvp` that looks like:

```
shape	color
circle	red|green|purple|orange|yellow|blue
square	orange|red|yellow|blue|purple|green
triangle	yellow|blue|red|purple|orange|green
```

The following works fine if colors are `;`-separated but I'm not sure how to specify a different FS.

```shell
mlr --csvlite --fs tab \
	nest --explode --values --across-records -f color  \
	color-shapes-nested.tsv
```

**Comment by johnkerl** [comment]
```
$ cat mydata.tsv
shape   color
circle  red|green|purple|orange|yellow|blue
square  orange|red|yellow|blue|purple|green
triangle        yellow|blue|red|purple|orange|green

$ mlr --csvlite --fs tab nest --explode --values --across-records -f color --nested-fs '|' mydata.tsv
shape   color
circle  red
circle  green
circle  purple
circle  orange
circle  yellow
circle  blue
square  orange
square  red
square  yellow
square  blue
square  purple
square  green
triangle        yellow
triangle        blue
triangle        red
triangle        purple
triangle        orange
triangle        green

$ mlr --csvlite --fs tab nest --explode --values --across-records -f color --nested-fs pipe mydata.tsv
shape   color
circle  red
circle  green
circle  purple
circle  orange
circle  yellow
circle  blue
square  orange
square  red
square  yellow
square  blue
square  purple
square  green
triangle        yellow
triangle        blue
triangle        red
triangle        purple
triangle        orange
triangle        green
```
My bad: I had inadvertently failed to invoke the part that replaces `pipe` with `|`. You should get it if you re-pull from head.

**Comment by johnkerl** [comment]
Also, I'm still working on some logic which explodes across fields by putting the exploded fields where the original was, rather than at the end of the record.

**Comment by johnkerl** [comment]
OK @aaronwolen , done with the in-place mods. This is currently in head:

Input:
``` 
$ cat w.tsv
a b
rs10  rs2237570,0.3,-0.72;rs117465896,0.75,-0.88
rs10041592  rs1628466,0.21,0.92;rs1729036,0.24,1;rs79177551,0.2,0.85
rs10033329  rs4835294,0.2,0.46
```

First step:
```
$ mlr --csvlite --fs tab nest --explode --values --across-records -f b w.tsv
a b
rs10  rs2237570,0.3,-0.72
rs10  rs117465896,0.75,-0.88
rs10041592  rs1628466,0.21,0.92
rs10041592  rs1729036,0.24,1
rs10041592  rs79177551,0.2,0.85
rs10033329  rs4835294,0.2,0.46
````

Second step:
```
$ mlr --csvlite --fs tab nest --explode --values --across-records -f b
  then nest --nested-fs comma --explode --values --across-fields -f b w.tsv
a b_1 b_2 b_3
rs10  rs2237570 0.3 -0.72
rs10  rs117465896 0.75  -0.88
rs10041592  rs1628466 0.21  0.92
rs10041592  rs1729036 0.24  1
rs10041592  rs79177551  0.2 0.85
rs10033329  rs4835294 0.2 0.46
```
along with:
```
$ mlr nest -h
Usage: mlr nest [options]
Explodes specified field values into separate fields/records, or reverses this.
Options:
  --explode,--implode   One is required.
  --values,--pairs      One is required.
  --across-records,--across-fields One is required.
  -f {field name}       Required.
  --nested-fs {string}  Defaults to ";". Field separator for nested values.
  --nested-ps {string}  Defaults to ":". Pair separator for nested key-value pairs.
Please use "mlr --usage-separator-options" for information on specifying separators.

Examples:

  mlr nest --explode --values --across-records -f x
  with input record "x=a;b;c,y=d" produces output records
    "x=a,y=d"
    "x=b,y=d"
    "x=c,y=d"
  Use --implode to do the reverse.

  mlr nest --explode --values --across-fields -f x
  with input record "x=a;b;c,y=d" produces output records
    "x_1=a,x_2=b,x_3=c,y=d"
  Use --implode to do the reverse.

  mlr nest --explode --pairs --across-records -f x
  with input record "x=a:1;b:2;c:3,y=d" produces output records
    "a=1,y=d"
    "b=2,y=d"
    "c=3,y=d"

  mlr nest --explode --pairs --across-fields -f x
  with input record "x=a:1;b:2;c:3,y=d" produces output records
    "a=1,b=2,c=3,y=d"

Notes:
* With --pairs, --implode doesn't make sense since the original field name has
  been lost.
* The combination "--implode --values --across-records" is non-streaming:
  no output records are produced until all input records have been read. In
  particular, this means it won't work in tail -f contexts. But all other flag
  combinations result in streaming (tail -f friendly) data processing.
* It's up to you to ensure that the nested-fs is distinct from your data's IFS:
  e.g. by default the former is semicolon and the latter is comma.
```

**Comment by aaronwolen** [comment]
Not sure how I missed this but I've been trying out the latest version and the nest functionality works perfectly. This has really simplified my workflow. Thanks, @johnkerl!

**Comment by johnkerl** [comment]
Back at you @aaronwolen -- thanks for helping make Miller better with some great feature requests!!

**Comment by severach** [comment]
Can we get another quoting option "--quote-original". Better spreadsheets detect quotes around numbers and import them as text instead of numbers without lots of hassle. Quote original keeps the same quoting as the source document.

Starting with
"FOO,BAR","1:30",1:40,1.50,"0616685"

Should not result in any of the following

"FOO,BAR",1:30,1:40,1.50,0616685
1:30 might be converted to a time when it should be text
0616685 will become numeric and lose the leading zero

"FOO,BAR","1:30","1:40","1.50","0616685"
1:40 won't be converted to a time
1.50 will become text and won't behave as a dollar amount should

With only 4 columns I can manually specify text. I can't with hundreds of columns.

**Comment by johnkerl** [comment]
OK. There are two issues here: conversion or non-conversion of strings to other things (like integers), and preservation of double-quotes.

For the latter issue, I can think about putting a was-quoted bit in the stream-record fields. Then, for fields which were present in the input data, quote them if that bit is set and if `--quote-original` is specified on the command line. For fields not present in the input data -- e.g. those computed using the `put` DSL -- I'm less interested in OR'ing or AND'ing the was-quoted bits of the fields. E.g. with `mlr put '$c = $a . $b'` if field `a` was quoted and `b` wasn't, should `c` be quoted or not?

For the former issue, Miller is already quite insistent on preserving text when it can. Verbs in general don't convert to number unless they need to -- e.g. `stats1 -f` fields, or `sort -n`. The exceptions are `filter` and `put` which by default do type-inferencing: for input `x=123`, field `x` will be parsed as integer. But this can be suppressed using `put -S` and `filter -S`, where (for example) field `x` would have *string* value `"123"` and you can cast using `int($x)` on your own choosing.

@severach can you give me some example Miller commands in your use case(s), to help me see which of these two (or both) are at play here?

**Comment by severach** [comment]
My report generator made a CSV file about 164MB. No spreadsheet can handle a file this large. My plan was to extract with mlr about 6 of the 50 or so fields of interest then use the spreadsheet to trim out unwanted records. This plan was halted when the quotes were removed from some of the numeric fields.

At 6 fields I could have told the spreadsheet which were numeric and which were text but this wouldn't be practical for a lot more fields or files sent to others who aren't aware of the need for keeping data types proper.

**Comment by johnkerl** [comment]
Did you try `--quote-all` ? And, can you copy/paste any miller commands you
tried?

On Tue, Jun 21, 2016 at 12:43 PM, Chris Severance <notifications@github.com>
wrote:

> My report generator made a CSV file about 164MB. No spreadsheet can handle
> a file this large. My plan was to extract with mlr about 6 of the 50 or so
> fields of interest then use the spreadsheet to trim out unwanted records.
> This plan was halted when the quotes were removed from some of the numeric
> fields.
>
> At 6 fields I could have told the spreadsheet which were numeric and which
> were text but this wouldn't be practical for a lot more fields or files
> sent to others who aren't aware of the need for keeping data types proper.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/johnkerl/miller/issues/77#issuecomment-227499330>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AB6m2nx4XthDXTYejaO9eOMX-i-Rfp1Tks5qOBS9gaJpZM4GQ0V_>
> .
>



-- 
John Kerl
kerl.john.r@gmail.com
http://johnkerl.org
https://github.com/johnkerl/miller

**Comment by severach** [comment]
--quote-all is no good as that would put quotes on the numeric values. Quotes on dollar amounts make them import as text. They don't justify right or add properly.

The idea here is that the spreadsheet recognizes quotes and always imports quoted text as text no matter what the contents and always imported unquoted text as money, counts, times, dates, and other numeric fields, only importing as text when all attempts at at recognizing some type of number have been exhausted. The generator has been designed to take advantage of this feature and has put quotes on the fields that need to be text and leaves the quotes out on fields that need to be numbers.

Imagine a field that contains both text and numbers, like a facility code.
A01
005
B01
006

Leaving the

**Comment by johnkerl** [comment]
OK from "My plan was to extract with mlr about 6 of the 50 or so fields of interest ..." the command I'm guessing is `mlr cut`. Sounds like computing fields is a non-issue since you're just cutting out the ones you need so we don't need to go down that road. Then a `--quote-original` combined with per-field was-quoted-on-input flag would suffice to solve your problem.

This sounds pretty easy -- not this week because RL but maybe next week I can draft something up -- ?

**Comment by johnkerl** [comment]
... quicker than I thought. This is head. Can you try it out & let me know?

```
$ cat reg_test/input/quote-original.csv 
a,b,c
1,2,3
4,"5",6
"7",8,"9"

$ mlr --csv --rs lf cut -o -f c,b,a reg_test/input/quote-original.csv 
c,b,a
3,2,1
6,5,4
9,8,7

$ mlr --csv --rs lf --quote-original cut -o -f c,b,a reg_test/input/quote-original.csv 
c,b,a
3,2,1
6,"5",4
"9",8,"7"
```

**Comment by severach** [comment]
Looks good.

**Comment by johnkerl** [comment]
Thanks for the awesome feature request!!

FYI -- while this is already in head, it will also go out in 4.3.0, which will probably be in a few weeks.

**Comment by sjackman** [comment]
Autodetect line endings! I never use CRLF line endings myself. Workaround: `alias mlr='--rs lf'`

**Comment by sjackman** [comment]
Add option `--tsv` as an alias for `--csv --fs tab`. I'd suggest also including `--rs lf` in the defaults for `--tsv`. I feel TSV is most commonly used on Unix, and unlike CSV RFC-4180, the IANA spec for `text/tab-separated-values` doesn't specify the line ending.

**Comment by sjackman** [comment]
It's easy to display the maximum value of a column `stats1 -a max -f foo`. Is there an easier way to display all the rows that have that maximum value for that column?
```sh
max=`mlr --csv --fs tab --onidx stats1 -a max -f NG50 stats.tsv`
mlr --csv --fs tab filter '$NG50 == '$max stats.tsv
```

**Comment by sjackman** [comment]
`--md` for a Markdown formatted table, which is pipe-separated-values (i.e. `--fs '|'`) and a bonus row separating the header and data, which is composed entirely of hyphens and pipes. See https://help.github.com/articles/organizing-information-with-tables/
I'm primarily interested in the output format to convert TSV to Markdown, but handling input as well would be nice.

**Comment by johnkerl** [comment]
ooh nice! agreed in & out would be nice, & out only is simpler (&
sufficient for you). i'll file a separate issue to track.

On Tue, Jun 28, 2016 at 7:19 PM, Shaun Jackman <notifications@github.com>
wrote:

> --md for a Markdown formatted table, which is pipe-separated-values (i.e. --fs
> '|') and a bonus row separating the header and data, which is composed
> entirely of hyphens and pipes. See
> https://help.github.com/articles/organizing-information-with-tables/
> I'm primarily interested in the output format to convert TSV to Markdown,
> but handling input as well would be nice.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/johnkerl/miller/issues/77#issuecomment-229213051>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AB6m2t5k8xJ5z9Yhpw-VxBg3XNS84LPGks5qQawEgaJpZM4GQ0V_>
> .
>



-- 
John Kerl
kerl.john.r@gmail.com
http://johnkerl.org
https://github.com/johnkerl/miller

**Comment by johnkerl** [comment]
@sjackman re max -- yes there's a better way to do two-pass using out-of-stream variables and `end` block. i'll post an example soon, but 1st, just acknowledging your post. :)

**Comment by johnkerl** [comment]
@sjackman no need for a new `emitx`; my bad. Here's an example:

```
$ cat myfile.tsv 
a   b   i   x   y
pan pan 1   2   0.989945
eks pan 2   1   0.775159
wye wye 3   1   0.761642
eks wye 4   5   0.322934
wye pan 5   5   0.448283
zee pan 6   3   0.323831
eks zee 7   5   0.554439
zee wye 8   5   0.400662
hat wye 9   4   0.763756
pan wye 10  4   0.498709
pan pan 1   3   0.474926
eks pan 2   1   0.165322
wye wye 3   0   0.074046
eks wye 4   2   0.913809
wye pan 5   5   0.217785
zee pan 6   4   0.838769
eks zee 7   3   0.590072
zee wye 8   1   0.363831
hat wye 9   5   0.595050
pan wye 10  1   0.127385
pan pan 1   2   0.946379
eks pan 2   3   0.775044
wye wye 3   1   0.718341
eks wye 4   5   0.324481
wye pan 5   2   0.367450
zee pan 6   2   0.760260
eks zee 7   1   0.784303
zee wye 8   2   0.815579
hat wye 9   3   0.529286
pan wye 10  2   0.051304
pan pan 1   2   0.161925
eks pan 2   1   0.948469
wye wye 3   2   0.209884
eks wye 4   0   0.412046
wye pan 5   5   0.169497
zee pan 6   3   0.740590
eks zee 7   5   0.257726
zee wye 8   4   0.210518
hat wye 9   3   0.083564
pan wye 10  4   0.190035
pan pan 1   4   0.318641
eks pan 2   2   0.122774
wye wye 3   1   0.152561
eks wye 4   5   0.033214
wye pan 5   5   0.254975
zee pan 6   4   0.306178
eks zee 7   3   0.833384
zee wye 8   2   0.136543
hat wye 9   3   0.239049
pan wye 10  2   0.702515
```

```
$ cat maxrows.sh 
mlr --csvlite --fs tab put -q '
  @records[NR] = $*;
  @maxx = max(@maxx,$x);
  end {
    for ((nr,k),v in @records) {
      if (k == "x") {
        if (@records[nr]["x"] == @maxx) {
          emit @records[nr]
        }
      }
    }
  }
' myfile.tsv
```

```
$ sh maxrows.sh 
a   b   i   x   y
eks wye 4   5   0.322934
wye pan 5   5   0.448283
eks zee 7   5   0.554439
zee wye 8   5   0.400662
wye pan 5   5   0.217785
hat wye 9   5   0.595050
eks wye 4   5   0.324481
wye pan 5   5   0.169497
eks zee 7   5   0.257726
eks wye 4   5   0.033214
wye pan 5   5   0.254975
```

The idea is to use an out-of-stream variable keyed by `NR` to keep all records, then loop over them at the end and conditionally emit them. The `if (k == "x") ...` part is awkward. `:^/`

**Comment by johnkerl** [comment]
markdown task https://github.com/johnkerl/miller/issues/106

**Comment by johnkerl** [comment]
@sjackman re autodetect, I definitely feel your pain ... but I lean away from that since it is hard to be efficient when reading from stdin.

In 4.3.0 there will be `export MLR_CSV_DEFAULT_RS=lf` (see also http://johnkerl.org/miller/doc/faq.html#Number_one_FAQ) which I think will help.

Re `--tsv` I've thought about it on & off & your ask tips me over. Coming soon! `:)`

**Comment by johnkerl** [comment]
`--tsv` is in head: https://github.com/johnkerl/miller/commit/190c92f567c4fac0765c1b75c10c790c34e83b0f

**Comment by johnkerl** [comment]
`--omd` is in head: https://github.com/johnkerl/miller/commit/fab0f513fe40026650aa4c73b2e5ed15d312d39e and some prior commits as well

@sjackman thank you for the great suggestions!!

**Comment by sjackman** [comment]
Thanks for your quick responses, John! For `--tsv` is the default record separator `--rs lf` or `--rs crlf`? Please say it's `--rs lf`!

**Comment by johnkerl** [comment]
... it is if you do a one-time setup of an environment variable:

http://johnkerl.org/miller/doc/faq.html#Number_one_FAQ

This way everyone can choose their default.

**Comment by sjackman** [comment]
Using an environment variable that is set in e.g. a `.zshrc` is not good for writing reproducible scripts that will behave identically when run on other machines that don't have that environment variable set. Good for interactive use though.

**Comment by johnkerl** [comment]
Good point ...

**Comment by sjackman** [comment]
I understand your reasoning for making CRLF the default line ending for CSV (namely that unfortunate RFC). For TSV (and really any non-CSV format) I'd suggest making the default line ending LF.

**Comment by johnkerl** [comment]
I get that. In Miller, the only difference between CSV and TSV is the field separator.

If you don't need double-quoting support, and if your fields don't have tabs or newlines in them, then perhaps you should use `--tsvlite`. That has LF as default RS and runs a bit faster as well.

**Comment by sjackman** [comment]
Oh, interesting. Thanks for the clarification. It sounds as though I've wanted `--csvlite` and `--tsvlite` all along. Does `--tsv` then treat two consecutive tabs as a literal tab character? That's quite unusual for TSV if so. There is unfortunately no standard way that I know of to represent a tab character in TSV. If I were to design the format, I'd suggest enclosing the field in double quotes, and using `\t` for tab like in C, but then we'd have one more non-standard TSV format. c.f. https://xkcd.com/927/

**Comment by johnkerl** [comment]
No, two consecutive tabs in a TSV file enclose an empty-string field, just as two consecutive commas in a CSV file enclose an empty-string field.

And if you use `--tsv` then, quite right, you can put a literal tab character inside double quotes since that's what the CSV spec supports, and Miller's TSV is CSV-but-with-tabs-instead-of-commas.

**Comment by sjackman** [comment]
Right. Thanks again, John. I've only just started using `mlr`, but it's already been very useful. Thanks!

**Comment by sjackman** [comment]
Are you familiar [ASCII delimited text](https://en.wikipedia.org/wiki/Delimiter#ASCII_delimited_text). It uses the standard ASCII character 31 (unit separator) to separate fields, ASCII character 30 (record separator) to separate records, ASCII character 29 (group separator) to separate tables, and ASCII character 28 (file separator) to separate files. I've never seen this format in the wild, but it would resolve once and for all the question of how to escape separator characters.

**Comment by johnkerl** [comment]
@sjackman I hadn't heard of that particular spec, but it's for that kind of thing that I allowed octal/hex support for Miller record-separators.

So you can already do that by
```
mlr --csvlite --rs '\x1d' --fs '\x1e' ...
```

We could also make that a bit less keystrokey by making a `--asv` option as a shorthand for the above.

**Comment by sjackman** [comment]
Cool! Probably no need for the option. Like I said, I've never seen it in the wild, and I probably won't use it until it has support from a text editor so that I can view and edit the files. A Vim plugin would probably do the trick though. If I ever find a serious use case for it, I'll make a comment here.

**Comment by johnkerl** [comment]
derp; thx!

**Comment by sjackman** [comment]
Does `mlr` have a command equivalent to `paste`?

**Comment by johnkerl** [comment]
@sjackman no it does not. It should be pretty easy to add, though.

(There's a concern about duplicate field names, e.g. if both left and right have a field named `x`. That can be solved via (a) caveat emptor, or (b) have optional left-prefix and right-prefix flags, e.g. `x` becomes `left_x` and `right_x`.)

That said, it becomes a little awkward within Miller's ARGF-stye file-processing semantics. Namely, given `mlr paste file1 file2 file3` one would want the three files pasted side by side. However Miller's ARGF-style processing means that -- for all other verbs -- there is a record stream taken from file1 + file2 + file3. So (in analogy with `mlr join`) we'd need something like `mlr paste -f file1 -f file2 file3`. That's a bit funny-looking since it breaks the symmetry of the filenames. But it would work.

**Comment by johnkerl** [comment]
The poor-man's way is to use the system `paste -d,` along with Miller's DKVP or CSV format. Or `paste -d' '` with PPRINT format.

**Comment by sjackman** [comment]
Thanks for the explanation, John. I used `paste` with TSV, which worked fine. Perhaps `mlr paste` could be implemented as a special case of `mlr join` where it's joining on the record index, `NR`.

**Comment by elyscape** [comment]
It would be great if `mlr` had a Bash completion script.

**Comment by torbiak** [comment]
I'm trying to reorder the columns of a histogram based on their sums so I can easily plot the top N series using a graphing package like gnuplot by using the leftmost N columns.

Eg, from this:

```
bin,a,b,c
0,1,3,2
```

To this:

```
bin,b,c,a
0,3,2,1
```

In a general purpose language I'd do something like a Schwartzian transform to get the order of the columns and then emit the binned data according to that order. The best I've come up with using miller is using command substitution and doing another pass to figure out the order and give it to the reorder verb:

```
echo -e 'a=1,b=3\na=2,b=4' >data
order=$(mlr --from data \
    stats1 -a sum --fx '' then \
    rename -r '(.*)_sum,\1' then \
    put -q 'for (k, v in $*) { emit {"col": k, "sum": v} }' then \
    sort -nr sum then \
    put -q '@order[$col] = 1; end { print joink(@order, ",") }')
mlr --from data reorder -f "$order"
```

Is there a better way? Am I missing something? I notice there's no sort function in the `put` DSL, which makes sense since there isn't a list datatype and sorting map keys seems really hacky. I guess I'm asking for a nicer way to reorder columns based on their aggregates, if it doesn't already exist and could fit within miller's design constraints.

**Comment by johnkerl** [comment]
@torbiak there's no `sort` function in the DSL but I've thought about it.  There would be sort by key, sort by value, and sort by function name (maybe a user-defined function). (In a 'real' language there would be sort-by-lambda/closure but that's a bit beyond the Miller DSL's simple syntax.)

You could use a Schwartzian transform to add fields in `put`, use `then`-chaining to feed into Miller `sort`, then back into `put` -- if this were along rows. Since it's along columns, no, probably not easily done in Miller.

FWIW I use this in Ruby for handling Miller-style DKVP lines:

```
def dkvpline2map(line)
  map = {}
  line.split(',').each do |pair|
    (k, v) = pair.split('=', 2)
    map[k] = v
  end
  map
end
```

**Comment by NikosAlexandris** [comment]
Dear miller-s,

# A

I have hundreds of files like

```
cat sample_median.csv
1,26.146306,1397700.000000,1553,1.04%
2,29.004897,198000.000000,220,0.15%
10,29.725383,19211400.000000,21346,14.24%
15,28.844068,230400.000000,256,0.17%
20,25.152257,91916100.000000,102129,68.12%
25,27.791520,696600.000000,774,0.52%
30,31.538513,6951600.000000,7724,5.15%
35,31.400053,774000.000000,860,0.57%
40,26.498211,8782200.000000,9758,6.51%
41,31.016598,133200.000000,148,0.10%
45,29.897872,593100.000000,659,0.44%
50,32.036074,4068000.000000,4520,3.01%
```

Via 
```
for ITEM in *categories.csv ;do
    mlr --csv label Code,Median_Temperature,Km,Pixels,Cover then put '$Cover = gsub($Cover, "%", "")' then put '$Weighted_Average = ($Cover * $Median_Temperature)/100' $ITEM
done
```
I get to
```
cat sample_median_wavg.csv
Code,Median_Temperature,Km,Pixels,Cover,Weighted_Average
2,29.004897,198000.000000,220,0.15,0.043507
10,29.725383,19211400.000000,21346,14.24,4.232895
15,28.844068,230400.000000,256,0.17,0.049035
20,25.152257,91916100.000000,102129,68.12,17.133717
25,27.791520,696600.000000,774,0.52,0.144516
30,31.538513,6951600.000000,7724,5.15,1.624233
35,31.400053,774000.000000,860,0.57,0.178980
40,26.498211,8782200.000000,9758,6.51,1.725034
41,31.016598,133200.000000,148,0.10,0.031017
45,29.897872,593100.000000,659,0.44,0.131551
50,32.036074,4068000.000000,4520,3.01,0.964286
```

Another example,
```
cat Acireale_100kmN16E47_06_median.csv
1,31.274191,50400.000000,56,0.13%
2,41.931690,182700.000000,203,0.46%
10,40.668863,1161000.000000,1290,2.91%
15,39.706989,78300.000000,87,0.20%
20,39.013383,9352800.000000,10392,23.42%
25,39.770772,135900.000000,151,0.34%
30,42.229262,2846700.000000,3163,7.13%
35,41.682468,840600.000000,934,2.10%
40,40.572099,21261600.000000,23624,53.24%
41,40.935475,72000.000000,80,0.18%
45,41.080757,642600.000000,714,1.61%
50,41.861756,3323700.000000,3693,8.32%
```
 
for which I get via 
```
mlr --csv label A,B,C,D,E then put '$E = gsub($E, "%", "")' then put '$WAvg = ($E * $B)/100' then stats1 -a sum -f WAvg Acireale_100kmN16E47_06_median.csv
```
to the sum of `Weighted_Average`
```
WAvg_sum
40.432765
```

I get further, in getting the result of
```
mlr --csv label A,B,C,D,E then put '$E = gsub($E, "%", "")' then put '$WAvg = ($E * $B)/100' then put -q '$A==50 {@ct=$B}; @sum += $WAvg; @index=@ct - @wsum; end{emitf @ct, @wsum, @index}' Acireale_100kmN16E47_06_median.csv
ct,sum,index
41.861756,40.432765,1.428991
```

This is, more or less, where I want to drive the data.

# B

I managed to join, the following two,
```
mlr --csv --rs lf cat Acireale_100kmN16E47_06_median.csv Acireale_100kmN16E47_07_median.csv
1,31.274191,50400.000000,56,0.13%
2,41.931690,182700.000000,203,0.46%
10,40.668863,1161000.000000,1290,2.91%
15,39.706989,78300.000000,87,0.20%
20,39.013383,9352800.000000,10392,23.42%
25,39.770772,135900.000000,151,0.34%
30,42.229262,2846700.000000,3163,7.13%
35,41.682468,840600.000000,934,2.10%
40,40.572099,21261600.000000,23624,53.24%
41,40.935475,72000.000000,80,0.18%
45,41.080757,642600.000000,714,1.61%
50,41.861756,3323700.000000,3693,8.32%

1,33.792107,50400.000000,56,0.13%
2,44.812347,182700.000000,203,0.46%
10,44.846281,1161000.000000,1290,2.91%
15,43.885570,78300.000000,87,0.20%
20,42.439296,9352800.000000,10392,23.42%
25,43.062736,135900.000000,151,0.34%
30,45.339263,2846700.000000,3163,7.13%
35,44.588770,840600.000000,934,2.10%
40,43.799767,21261600.000000,23624,53.24%
41,43.425986,72000.000000,80,0.18%
45,44.251293,642600.000000,714,1.61%
50,44.859712,3323700.000000,3693,8.32%
```
like
```
mlr --implicit-csv-header --nidx --csv join -f Acireale_100kmN16E47_06_median.csv -j 1 -r 1 --rp "07_" then cut -o -f 1,2,07_2,3,4 Acireale_100kmN16E47_07_median.csv
1,2,07_2,3,4
1,31.274191,33.792107,50400.000000,56
2,41.931690,44.812347,182700.000000,203
10,40.668863,44.846281,1161000.000000,1290
15,39.706989,43.885570,78300.000000,87
20,39.013383,42.439296,9352800.000000,10392
25,39.770772,43.062736,135900.000000,151
30,42.229262,45.339263,2846700.000000,3163
35,41.682468,44.588770,840600.000000,934
40,40.572099,43.799767,21261600.000000,23624
41,40.935475,43.425986,72000.000000,80
45,41.080757,44.251293,642600.000000,714
50,41.861756,44.859712,3323700.000000,3693
```
and get some math I need like this:
```
mlr --implicit-csv-header --nidx --csv --opprint join -f Acireale_100kmN16E47_06_median.csv -j 1 -r 1 --rp "07_" then cut -o -f 1,2,07_2,3,4,5 then put '$5 = gsub($5, "%", "")' then put '$WAvg = ($5 * (($2 + $07_2) / 2) )/100' Acireale_100kmN16E47_07_median.csv
1  2         07_2      3               4     5     WAvg
1  31.274191 33.792107 50400.000000    56    0.13  0.042293
2  41.931690 44.812347 182700.000000   203   0.46  0.199511
10 40.668863 44.846281 1161000.000000  1290  2.91  1.244245
15 39.706989 43.885570 78300.000000    87    0.20  0.083593
20 39.013383 42.439296 9352800.000000  10392 23.42 9.538109
25 39.770772 43.062736 135900.000000   151   0.34  0.140817
30 42.229262 45.339263 2846700.000000  3163  7.13  3.121818
35 41.682468 44.588770 840600.000000   934   2.10  0.905848
40 40.572099 43.799767 21261600.000000 23624 53.24 22.459791
41 40.935475 43.425986 72000.000000    80    0.18  0.075925
45 41.080757 44.251293 642600.000000   714   1.61  0.686923
50 41.861756 44.859712 3323700.000000  3693  8.32  3.607613
```

And, finally, **the question is**: can I combine # A and # B, for three, identically structured, files, in one go?  Then, loop over hundreds of similar files, placed in independent directories? A small push would help me greatly.

**Comment by NikosAlexandris** [comment]
I have progressed using csvjoin/csvcut first, like
```
function statistics {
    csvjoin -c 1 -H ${@} |csvcut -c 1,2,6,10,3,4,5 |mlr --csv --opprint label Code,June,July,August,Sq.km.,Pixels,Cover then put '$Average = ($June + $July + $August)/3' then put '$Weighted_Average = $Cover * $Average / 100' ;
}
```

It would be nice to have, if there isn't, such a join functionality in (the awesome!) miller toolkit.

**Comment by singalen** [comment]
How do I `tail -f` after `--csv cut -f`?
To be efficient, one only needs to parse the first line of a file, and then go into `tail -f` mode.
Can one do this now?
Another question, can I select fields for `cut` with mask/regex/prefix?

**Comment by johnkerl** [comment]
@NikosAlexandris thanks! I haven't replied -- need some time to digest but am backlogged by RL this week. :(

**Comment by NikosAlexandris** [comment]
Further, how can I deduplicate the following code, that lives within a shell script?

```
function average_two_wrspr_June() {
    csvjoin -H -c 1 "$@" --outer |mlr --csv label Code,June1,Area1,Pixels1,Cover1,CodeR,June2,Area2,Pixels2,Cover2 then cut -f Code,June1,Area1,Pixels1,Cover1,June2,Area2,Pixels2,Cover2 then put '$June = ( (is_not_null($June1) ? $June1 : 0.0) * (is_not_null($Pixels1) ? $Pixels1 : 0) + (is_not_null($June2) ? $June2 : 0.0) * (is_not_null($Pixels2) ? $Pixels2 : 0) ) / ( (is_not_null($Pixels1) ? $Pixels1 : 0) + (is_not_null($Pixels2) ? $Pixels2 : 0) )' then put '$Area = (is_not_null($Area1) ? $Area1 : 0) + (is_not_null($Area2) ? $Area2 : 0)' then put '$Pixels = (is_not_null($Pixels1) ? $Pixels1 : 0) + (is_not_null($Pixels2) ? $Pixels2 : 0)' then put '$Cover = (is_not_null($Cover1) ? $Cover1 : 0) + (is_not_null($Cover2) ? $Cover2 : 0)' then cut -f Code,June,Area,Pixels,Cover |sed '1d' ;
}

function average_two_wrspr_July() {
    csvjoin -H -c 1 "$@" --outer |mlr --csv label Code,July1,Area1,Pixels1,Cover1,CodeR,July2,Area2,Pixels2,Cover2 then cut -f Code,July1,Area1,Pixels1,Cover1,July2,Area2,Pixels2,Cover2 then put '$July = ( (is_not_null($July1) ? $July1 : 0.0) * (is_not_null($Pixels1) ? $Pixels1 : 0) + (is_not_null($July2) ? $July2 : 0.0) * (is_not_null($Pixels2) ? $Pixels2 : 0) ) / ( (is_not_null($Pixels1) ? $Pixels1 : 0) + (is_not_null($Pixels2) ? $Pixels2 : 0) )' then put '$Area = (is_not_null($Area1) ? $Area1 : 0) + (is_not_null($Area2) ? $Area2 : 0)' then put '$Pixels = (is_not_null($Pixels1) ? $Pixels1 : 0) + (is_not_null($Pixels2) ? $Pixels2 : 0)' then put '$Cover = (is_not_null($Cover1) ? $Cover1 : 0) + (is_not_null($Cover2) ? $Cover2 : 0)' then cut -f Code,July,Area,Pixels,Cover |sed '1d' ;
}
```

The difference above, to get from the first to the second, is only `s/June/July/`.  I have tried to use positional parameter(s), i.e. in place of 'June' and 'July', use "${1}" in one generic function. And, replacing `"$@"` with `"${@:2}"`.

ps- Apologies for the multiple posts here.

**Comment by johnkerl** [comment]
@NikosAlexandris unless I misunderstand, the June/July problem can be solved by

```
function average_two_wrspr_month() {
  month_name=$1
  csvjoin -H -c 1 "$@" --outer \
  | mlr --csv label Code,${month_name}1,Area1,Pixels1,Cover1,CodeR,${month_name}2,Area2,Pixels2,Cover2 \
    then cut -f Code,${month_name}1,Area1,Pixels1,Cover1,${month_name}2,Area2,Pixels2,Cover2 \
    then put \
      '$'$month_name' = ( (is_not_null($'$month_name'1) ? $'$month_name'1 : 0.0) * (is_not_null($Pixels1) ? $Pixels1 : 0) + (is_not_null($'$month_name'2) ? $'$month_name'2 : 0.0) * (is_not_null($Pixels2) ? $Pixels2 : 0) ) / ( (is_not_null($Pixels1) ? $Pixels1 : 0) + (is_not_null($Pixels2) ? $Pixels2 : 0) )
      '
    then put '$Area = (is_not_null($Area1) ? $Area1 : 0) + (is_not_null($Area2) ? $Area2 : 0)' \
    then put '$Pixels = (is_not_null($Pixels1) ? $Pixels1 : 0) + (is_not_null($Pixels2) ? $Pixels2 : 0)' \
    then put '$Cover = (is_not_null($Cover1) ? $Cover1 : 0) + (is_not_null($Cover2) ? $Cover2 : 0)' \
    then cut -f Code,${month_name},Area,Pixels,Cover \
    | sed 1d
}
```

namely, take the month name as positional parameter, but outside the single-quoted `put` statements.

**Comment by NikosAlexandris** [comment]
Here, some subtle but necessary modification:
```
function average_two_wrspr_month() {
    month_name=$1
    csvjoin -H -c 1 "${@:2}" --outer \
        | mlr --csv label \
        Code,${month_name}1,Area1,Pixels1,Cover1,CodeR,${month_name}2,Area2,Pixels2,Cover2 \
    then cut -f \
        Code,${month_name}1,Area1,Pixels1,Cover1,${month_name}2,Area2,Pixels2,Cover2 \
    then put \
        '$'$month_name' = ( (is_not_null($'$month_name'1) ? $'$month_name'1 :0.0) * (is_not_null($Pixels1) ? $Pixels1 : 0) + (is_not_null($'$month_name'2) ? $'$month_name'2 : 0.0) * (is_not_null($Pixels2) ? $Pixels2 : 0) ) / ( (is_not_null($Pixels1) ? $Pixels1 : 0) + (is_not_null($Pixels2) ? $Pixels2 : 0) )' \
    then put \
        '$Area = (is_not_null($Area1) ? $Area1 : 0) + (is_not_null($Area2) ? $Area2 : 0)' \
    then put \
        '$Pixels = (is_not_null($Pixels1) ? $Pixels1 : 0) + (is_not_null($Pixels2) ? $Pixels2 : 0)' \
    then put \
        '$Cover = (is_not_null($Cover1) ? $Cover1 : 0) + (is_not_null($Cover2) ? $Cover2 : 0)' \
    then cut -f \
        Code,${month_name},Area,Pixels,Cover \
        | sed 1d
}
```

It works. Awesome!

By the way, I had to dig to get to [Null_data: empty and absent](http://johnkerl.org/miller/doc/reference.html#Null_data:_empty_and_absent) and the syntax to use. Empty values are frequently an issue to deal with when massaging data. Wouldn't it be useful to have a `-n` flag to `put`, with a numeric argument (`int` or `float`) to do this
```
(is_not_null($'$month_name'1) ? $'$month_name'1 :0.0)
```
automatically? Something like
```
mlr --csv cat then put -n Area1,0 -n Area2,0 '$Area = $Area1 + $Area2' file.csv
```

or
```
mlr --csv cat then put -n Area1,Area2 -v 0  '$Area = $Area1 + $Area2' file.csv
```
?

**Comment by pjfarleyiii** [comment]
Question 1: In Debian (buster, V10) the latest available miller from the stable repository is V5.4.0.  Can you say when any more recent releases will be available for debian 10?  I checked both testing and unstable repositories and saw only 5.4.0 there.

Question 2: Is the file mlr.linux.x86_64 in the 5.6.2 release assets area a binary executable that will run in debian 10?

Thanks for what looks like a very suitable replacement for the csvfix tool I have been using under Win10, I will be investigating it extensively over the next few weeks.

Regards,

Peter

**Comment by sjackman** [comment]
You can also install (and upgrade!) the most recent version of Miller on (nearly) any distribution of Linux using Homebrew on Linux: https://docs.brew.sh/linux

**Comment by pjfarleyiii** [comment]
Thanks for the homwbrew link, I will check that out.

**Comment by pjfarleyiii** [comment]
Wishlist # 1: Full documentation of regex syntax and semantics accepted by miller.  In particular, what "non standard" regex abbreviations are supported?  E.G., is the gawk extension "\s" abbreviation for [[:space:]] accepted?  Are PCRE abbreviations (\w, etc.) accepted?  Are POSIX character classes like [[:digit:]] accepted?  Current documentation is very sparse on this subject.

Alternative # 1: URL(s) to existing regex documentation that covers the miller implementation (e.g., libexpat, PCRE, etc.) plus documentation of any exceptions to those versions not supported or supported differently by miller.

Wishlist # 2: Much more concise table of DSL functions and operators plus summary tables "by type" (e.g., arithmetic, statistical, string functions and operators, etc).  The current web format is very, very long and provides no summary by type for quick reference.

**Comment by pjfarleyiii** [comment]
Wishlist: A new DSL function, strmatch(str, regex, array), modeled on the GNU awk extensions to the awk "match()" function, including the standard return value of the first position matched.  For a non-match condition, it would have to return -1 or other unique invalid position value since miller string position 1 is 0 while awk uses 1 for position 1.

The GNU awk extension of the third array argument would be particularly useful,  In gawk usage, this array contains the full matched string (array[0]) and each matched pattern group (array[groupno], 1 to # of groups), and array[groupno, "start"] contains the starting position of each matched (sub-)string and array[groupno, "length"] contains the length of that matched (sub-)string.

For ease of miller implementation and probable usage (most likely as argument #3 to a substr() function call) the [groupno, "length"] subscript could be [groupno, "endpos"] instead to match the definition of the miller substr() function third argument.

Non-preferred alternative implementation: A predefined reserved array name to contain the match() third argument output values after every successful "=~" match operation.

The purpose here is to provide actual variable names which contain the values of the matched (sub-)strings and their positions in the entire matched string, and not just "back reference" values (\1, \2, etc.) that can only be embedded in literal strings.

Obviously this can be provided as a user function assigning values to an internal map or array returned as the function value, but this is not as coder-friendly since it must then be specifically invoked after each "=~" operation.

Peter

**Comment by sjackman** [comment]
FWIW I mean big fan of R's tidyverse family of data manipulation functions. The string library is called `stringr`, and the function that returns positions of matches is `str_locate`:
https://www.rdocumentation.org/packages/stringr/versions/1.4.0
https://github.com/rstudio/cheatsheets/blob/master/strings.pdf
https://www.rdocumentation.org/packages/stringr/versions/1.4.0/topics/str_locate

**Comment by sjackman** [comment]
Feature request: Add `-f` option to `mlr unsparsify -f`. `-f` specifies the fields to unsparsify. The `unsparsify` operating is stream when `-f` is specified.

**Comment by johnkerl** [comment]
Hey Shaun long time no see! :) I hope all is well with you.

Do you mean: if `mlr unsparsify foo.dat` then operate on all columns; if `mlr unsparsify -f a,b,c foo.dat` then only add new fields if they're named one of `a`, `b`, or `c` -- ?

**Comment by sjackman** [comment]
> Hey Shaun long time no see! :) I hope all is well with you.

👋 Hi! All is well with me. Likewise I hope that all is well with you. =)

> Do you mean: if mlr unsparsify foo.dat then operate on all columns; if mlr unsparsify -f a,b,c foo.dat then only add new fields if they're named one of a, b, or c -- ?

Yes! And the primary motivation is to be able to stream `mlr unsparsify` for large datasets.

**Comment by johnkerl** [comment]
sweet :) -- #387 it is! :)

**Comment by sjackman** [comment]
A pretty typical Miller invocation for me looks like…
```console
$ pv foobar.tsv.gz | pigz -d | mlr -t count -f foo then sort -nr count then head
 396MiB 0:00:53 [7.41MiB/s] [================================>] 100%
foo count
bar 123
```

`pv` is a superuseful command to display a progress bar.
`pigz -d` decompresses in parallel faster than `gzip`.

`pv foobar.tsv.gz | mlr --prepipe "pigz -d" -t count -f foo then sort -nr count then head` works!

`mlr --prepipe 'pv | pigz -d' --csv count -g barcode 1084526_1091679/outs/probe_info.csv.gz` works in that it gives the correct answer, but the progress bar is not displayed, because `pv` doesn't know the file size unless it's given the name of the file as a command line argument _or_ the file (but not a pipe) is it's standard input.

So this does work, it gives the right answer *and* displays the progress bar:
`pv <1084526_1091679/outs/probe_info.csv.gz | mlr --prepipe 'pigz -d' --csv count -g barcode`

So on to my feature request.

1. When the input to `mlr` is coming from a file, could it open the `prepipe` process with `stdin` be that actual file, rather than a pipe?
2. Would you consider adding a `-P` option that display a progress bar, possibly by piping into `pv`, in which case it's just a shortcut for `--prepipe pv`?
3. Would you consider add a `-z` option to guess the appropriate decompress program to use based on the input filename or alternatively guessed based on the first few bytes of the file? (like `libmagic` and file`)

These two features would cut my example command down from…
`pv foobar.tsv.gz | pigz -d | mlr -t count -f foo then sort -nr count then head`
to
`mlr -Ptz count -f foo then sort -nr count then head foobar.tsv.gz`

**Comment by aborruso** [comment]
Hi @johnkerl it would be great to have a cut verb, based on filter.

Something like `mlr --csv cut -filter "match.+" input.csv`, to remove all columns that have at least one cell that match the regex `match.+`.

I know, it can be a long process, but for certain kind of data, filled with a lot of metadata header lines, it would be very comfortable.  Below an example in which I must cut the fields that contain `Aver`.

![image](https://user-images.githubusercontent.com/30607/120114091-cb406580-c17d-11eb-930e-e2d26078050d.png)

**Comment by NikosAlexandris** [comment]
To have a quick easy-to-read of the headers of a CSV, I do:
```
head -1 file.csv |tr ',' '\n' |cat -n
```
Isn't `mlr headers file.csv` something nice to have?

ps- `xsv headers` example in https://github.com/BurntSushi/xsv#a-whirlwind-tour.

**Comment by johnkerl** [comment]
@NikosAlexandris true.

`mlr headers` is great to want.

Miller doesn't do well with zero-data-record output.

I might be able to add a 'has-no-data' flag to the miller-record data structure, at some code complexity ...

**Comment by aborruso** [comment]
> To have a quick easy-to-read of the headers of a CSV, I do:

If you have

```
a,b
NA,3
4,NA
5,6
```

and run `head -1 input.csv | mlr --c2x --implicit-csv-header cat`, you will have

```
1 a
2 b
```

I know, it's not a verb :)

**Comment by NikosAlexandris** [comment]
> > To have a quick easy-to-read of the headers of a CSV, I do:
> 
> If you have
> 
> ```
> a,b
> NA,3
> 4,NA
> 5,6
> ```
> 
> and run `head -1 input.csv | mlr --c2x --implicit-csv-header cat`, you will have
> 
> ```
> 1 a
> 2 b
> ```
> 
> I know, it's not a verb :)

Indeed, my command line history is rich of `--c2x` already :-). But a `headers` like `xsv`'s (along with `-j`) would be nice to have. In case this is an easy addition.

**Comment by sjackman** [comment]
I use `mlr --c2x head -n 1 | cat -n` for this purpose myself.

**Comment by NikosAlexandris** [comment]
> mlr --c2x head -n 1 | cat -n

Me too. I did. But I prefer `xsv` and saving keystrokes. You already know, but here:
```
➜  mlr --c2x head -n 1 example.csv | cat -n
     1  x    99
     2  y    8
     3  e29
     4  e31
     5  e32
     6  mn
     7  ln
     8  le
     9  date 2019-12-25
    10  time 00:00:00
```
and
```
➜  xsv headers example.csv
1   x
2   y
3   e29
4   e31
5   e32
6   mn
7   ln
8   le
9   date
10  time
```
or even
```
➜  xsv headers -j example.csv
x
y
e29
e31
e32
mn
ln
le
date
time
```

**Comment by trantor** [comment]
Yep, something along the lines of [csvlook](https://csvkit.readthedocs.io/en/latest/scripts/csvlook.html) from csvkit would be nice too

**Comment by johnkerl** [comment]
Having this catch-all task open for people to drop in a quick comment, rather than creating a new issue, wasn't a bad idea. However I find that it gets low use, people are generally OK with opening new issues at https://github.com/johnkerl/miller/issues, and (worst) it's easy to 'miss' new comments here and fail to see them as asks.

So let's just open new issues at https://github.com/johnkerl/miller/issues so we can track things better.
